### PR TITLE
Fix for issue #4658: Add serialization performance tests for XmlSerializer and DCJS

### PR DESCRIPTION
**Pull request by SGuyGe**
@khdang @shmao @zhenlan

### CODE REVIEW

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.Serialization.Json;
+using System.Xml;
+
+namespace System.Runtime.Tests.Performance
+{
+    public class DcjsDeserializationTests
+    {
+        public void RunDcjsDeserializeTest(object obj, int iterations)
```
Comment (shmao) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: The class name uses "Deserialization" while this method name uses "Deserialize". Do we want to keep them consistent?

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            var serializer = new DataContractJsonSerializer(obj.GetType());
```
Comment (shmao) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: Question: will we run the Perf tests in Net Native? If so, will the above line work in Net Native?

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+            using (var stream = new MemoryStream())
+            {
+                serializer.WriteObject(stream, obj);
+                foreach (var iteration in Benchmark.Iterations)
+                    using (iteration.StartMeasurement())
```
Comment (shmao) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: I thinks it'd be better to use curly brackets for the above two statements for clarity .

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                        {
+                            stream.Position = 0;
+                            serializer.ReadObject(stream);
+                        }
+            }
+        }
+
+        public static Byte[] CreateByteArray(int size)
+        {
+            Byte[] obj = new byte[size];
```
Comment (shmao) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: Nit: Byte[] -> byte[]

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        public void DeserializeByteArrayOf1M()
+        {
+            byte[] byteArrayOf1M = CreateByteArray(1024 * 1024);
+            RunDcjsDeserializeTest(byteArrayOf1M, 10);
+        }
+
+        [Benchmark]
+        public void DeserializeStringOf128Byte()
+        {
+            string stringOf128Bytes = new string('a', 1024);
```
Comment (shmao) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: Should this be string('a', 128) ?

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                obj[i] = (byte)(i % 256);
+            }
+            return obj;
+        }
+
+        [Benchmark]
+        public void DeserializeByteArrayOf1K()
+        {
+            byte[] byteArrayOf1K = CreateByteArray(1024);
+            RunDcjsDeserializeTest(byteArrayOf1K, 10000);
```
Comment (shmao) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: Do we want to use a constant variable for the number "10000"?

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsSerializationTests.cs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        public void SerializeByteArrayOf1M()
+        {
+            byte[] byteArrayOf1M = CreateByteArray(1024 * 1024);
+            RunDcjsSerializeTest(byteArrayOf1M, 100);
+        }
+
+        [Benchmark]
+        public void SerializeStringOf128Byte()
+        {
+            string stringOf128Bytes = new string('a', 1024);
```
Comment (shmao) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsSerializationTests.cs]: Same as above. Should this be string('a', 128)?

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs (hunk 2 above)
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: No.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs (hunk 1 above)
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: Yeah, that would be better, updated.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs (hunk 3 above)
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: Agree, updated.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs (hunk 4 above)
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: Updated.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs (hunk 6 above)
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: Yes, that number varies and is picked according to individual tests.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs (hunk 5 above)
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: Good catch, fixed.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsSerializationTests.cs (hunk 7 above)
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsSerializationTests.cs]: Yes, fixed.

**Comment by stephentoub** [comment]
LGTM

**Comment by stephentoub** [comment]
cc: @ianhays

File: src/System.Xml.XmlSerializer/tests/Performance/XmlDeserializationTests.cs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        public void DeserializeByteArrayOf1M()
+        {
+            byte[] byteArrayOf1M = CreateByteArray(1024 * 1024);
+            RunXmlDeserializationTest(byteArrayOf1M, 10);
+        }
+
+        [Benchmark]
+        public void DeserializeStringOf128Byte()
+        {
+            string stringOf128Bytes = new string('a', 1024);
```
Comment (shmao) [inline src/System.Xml.XmlSerializer/tests/Performance/XmlDeserializationTests.cs]: should be string('a', 128);

File: src/System.Xml.XmlSerializer/tests/Performance/XmlDeserializationTests.cs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            RunXmlDeserializationTest(stringOf128Bytes, 10000);
+        }
+
+        [Benchmark]
+        public void DeserializeStringOf1024Byte()
+        {
+            string stringOf1024Bytes = new string('k', 128);
```
Comment (shmao) [inline src/System.Xml.XmlSerializer/tests/Performance/XmlDeserializationTests.cs]: should be string('a', 1024);

**Comment by shmao** [comment]
LGTM.

**Pull request closed without merging**